File: Ruby/problem014.rb
#! /usr/bin/env ruby
=begin
Problem 14 - Project Euler
http://projecteuler.net/index.php?section=problems&id=14
=end
$cache = {}

def chain n
    if $cache[n]
        return $cache[n]
    elsif n == 1
        return 1
    end

    collaz = if n.even?
        1 + chain(n / 2)
    else
        1 + chain(3 * n + 1)
    end
    $cache[n] = collaz
    collaz
end

maxlen = 0
maxval = 0
for i in 1..1_000_000
    c = chain i
    if c > maxlen
        maxlen = c
        maxval = i
    end
end
puts maxlen, maxval

File: Kotlin/problem014.kt
// Problem 14 - Project Euler
// http://projecteuler.net/index.php?section=problems&id=14
fun main(args: Array<String>) {
    var map = mutableMapOf(1L to 1L)
    for (i in 2L..1000000) {
        var x = i
        var count = 0L
        while (x != 1L) {
            val mx = map[x]
            if (mx != null) {
                count += mx
                break
            } else {
                x = f(x)
                count++
            }
        }
        map[i] = count
    }
    val maxitem = map.maxBy { it.value }
    println(maxitem?.key)
}

fun f(x: Long): Long {
    return if (x % 2 == 0L) x / 2 else 3 * x + 1
}
